### PR TITLE
[R-package] Disabled early stopping when using 'dart' boosting strategy

### PR DESCRIPTION
**Pull request by jameslamb**
See discussion in #1893 for background. In this PR, I propose disabling early stopping when the user chooses `'dart'` boosting. The same change was made in the Python package in #1895 .

This PR also adds the additional early stopping alias that was added in #2431 .

While working on this piece of code, I refactored it to make it a bit more obvious what is happening. I found that the same code is copied in `lgb.cv()` and `lgb.train()`. I think this duplication could lead to inconsistencies in the future, but I considered reducing that duplication out of scope.

### CODE REVIEW

**Comment by StrikerRUS** [comment]
@jameslamb Please see [this my comment](https://github.com/microsoft/LightGBM/issues/1893#issuecomment-476201045).

**Comment by jameslamb** [comment]
> @jameslamb Please see [this my comment](https://github.com/microsoft/LightGBM/issues/1893#issuecomment-476201045).

@StrikerRUS apologies, I actually meant to address that comment in PR description. What do you think should be done? Since users are just passing in functions to `callbacks`, I don't know how to tell whether those functions are for early stopping or some other purpose.

Are you just suggesting that I invoke this warning WHENEVER someone uses `dart` boosting, so that people who passed early stopping custom callbacks will know?

**Comment by StrikerRUS** [comment]
@jameslamb In my comment from the original issue I meant that this fix covers the case when user passes `boosting = 'dart'` and `early_stopping_rounds = x`. However, it seems that the fix doesn't cover the case when user passes `boosting = 'dart'` and `callbacks = list(cb.early.stop(x))`. Am I wrong? The Python-package fix covers the both cases. I think it's possible just to mimic the Python code fix or
> Since users are just passing in functions to callbacks, I don't know how to tell whether those functions are for early stopping or some other purpose.

https://github.com/microsoft/LightGBM/blob/1f88721a7982a804731b1dc108ee40625405639c/R-package/R/callback.R#L386

maybe somehow use this information?

**Comment by Laurae2** [comment]
@jameslamb @StrikerRUS From what I read, is it `any(unlist(lapply(callback, function(x) {attr(x, "name")})) == "cb.early.stop")` which should be also added as a condition?

**Comment by StrikerRUS** [comment]
@jameslamb @Laurae2 Maybe simply move the check into the callback function as it was done for Python-package? https://github.com/microsoft/LightGBM/pull/1895/files

**Comment by jameslamb** [comment]
> @jameslamb @StrikerRUS From what I read, is it `any(unlist(lapply(callbacks, function(x) {attr(x, "name")})) == "cb.early.stop")` which should be also added as a condition?

to your @StrikerRUS 's comment...yes I could add that and am happy to, but `cb.early.stop()` is not exported from the package (https://github.com/microsoft/LightGBM/blob/master/R-package/NAMESPACE). So we can add that check (or something similar) to 100% prevent "any early stopping callbacks that are part of the LightGBM package", but I don't think there is anything we can do about the case where some user writes their own totally-custom thing and passing it to `callbacks`. I think that's fine, just want to clarify the difference.

As far as pushing the logic down into `callback()` like #1895 , I personally could not figure out how to do that in the R package. The structure of `callback.R` is quite a bit different from `callback.py`, as far as I can tell. Let me take another look at `cb.early.stop()` and see if I can mimic the behavior.

**Comment by StrikerRUS** [comment]
@jameslamb 
> ... but cb.early.stop() is not exported from the package 

Ah, I didn't know that!
So, I added known issue 6 in https://github.com/microsoft/LightGBM/issues/1143#issuecomment-526936690 by a mistake, right?

> but I don't think there is anything we can do about the case where some user writes their own totally-custom thing and passing it to callbacks.

I don't think that someone would write a custom callback for early stopping, since we already have early stopping callback. Even if one will, they will use our implementation as a reference and will see the prohibition of using early stopping with the `dart` mode.

**Comment by jameslamb** [comment]
> @jameslamb
> 
> > ... but cb.early.stop() is not exported from the package
> 
> Ah, I didn't know that!
> So, I added known issue 6 in [#1143 (comment)](https://github.com/microsoft/LightGBM/issues/1143#issuecomment-526936690) by a mistake, right?
> 
> > but I don't think there is anything we can do about the case where some user writes their own totally-custom thing and passing it to callbacks.
> 
> I don't think that someone would write a custom callback for early stopping, since we already have early stopping callback. Even if one will, they will use our implementation as a reference and will see the prohibition of using early stopping with the `dart` mode.

I don't think that known issue was a mistake, maybe you just meant something different than what I thought. We do have a keyword argument `callbacks` in `lgb.train()` and `lgb.cv()`, so I assumed your comment was about more thoroughly documenting the expect form of functions passed to that argument. Right now the document just says "A list of callback functions to evaluate at each iteration"

So our approach right now is to say "you can enable early stopping by passing `early_stopping_rounds` and providing a validation set to `valids`", not "add function `cb.early_stop()` to `callbacks` argument". `callbacks` is a generic mechanism the R package uses for "any code you want to run after each iteration". By the way this is the closest I've looked at this section of the code in a while so @Laurae2 may correct me, but I think that is how it works.

I think for now it would be sufficient to add the check using `attr()` to see if `cb.early.stop` has added to the list in `callbacks`.

**Comment by StrikerRUS** [comment]
@jameslamb Nope, I thought right about documenting functions in `callback.R`, because I was sure that users are encouraged take them and pass to `callbacks` argument.

**Comment by jameslamb** [comment]
> @jameslamb Nope, I thought right about documenting functions in `callback.R`, because I was sure that users are encouraged take them and pass to `callbacks` argument.

Ok then yes, I misinterpreted what you meant. Documenting would also mean exporting. This is one of those things where R and Python are really different...in R you have to explicitly declare things as exported to make them usable. I'll update the language in #1143 to say "export callback functions" if that's alright.

**Comment by jameslamb** [comment]
Ok @Laurae2 @StrikerRUS whenever you have time, could you take another look? I've added protection against `cb.early.stop()` being passed directly via `callbacks` parameter.

I still left this in `lgb.train()` and `lgb.cv()` because I couldn't figure out how to do it all the way down in `callback()` like in the Python package.

**Comment by StrikerRUS** [comment]
@jameslamb 
> I'll update the language in #1143 to say "export callback functions" if that's alright.

I think that then this is another issue, even more feature request. But anyway, it's not related to pkgdown issue. I'll remove it from that list. Can you please create a separate issue for exporting callback functions?

File: R-package/R/lgb.cv.R [hunk 1]
```diff
@@ -211,18 +211,42 @@ lgb.cv <- function(params = list(),
...
+    first_early_stop_param <- which(early_stop_param_indx)[[1]]
+    first_early_stop_param_name <- names(params)[[first_early_stop_param]]
+    early_stopping_rounds <- params[[first_early_stop_param_name]]
+  }
+
+  # Did user pass parameters that indicate they want to use early stopping?
+  using_early_stopping_via_args <- !is.null(early_stopping_rounds)
+
+  # Cannot use early stopping with 'dart' boosting
+  if (identical(params$boosting, "dart")){
```
Comment (StrikerRUS) [inline R-package/R/lgb.cv.R]: Does it cover cases when `dart` is passed via [boosting aliases](https://lightgbm.readthedocs.io/en/latest/Parameters.html#boosting)?

Comment (jameslamb) [reply R-package/R/lgb.cv.R]: Seems we don't actually use boosting aliases in the R package at all currently. We will pass through anything you pass into `...` in `lgb.train()` or `lgb.cv()` but we have no code I see in the R package that explictly needs to access that value. I'll add them to this PR since it seems like it would be in scope.

Comment (jameslamb) [reply R-package/R/lgb.cv.R]: Ok I just pushed https://github.com/microsoft/LightGBM/pull/2443/commits/cab2bcdd011e85c4dda008e69acc2ac390b63674 which adds the boosting aliases. I also added a unit test in that commit which will check that the warning is generated for all the aliases.

I think this approach of centralizing the alias map is preferable to having the aliases hard-coded as string literals in multiple places. For now I only added the boosting-related ones as that's within the scope of this PR. Let me know what you think about the approach! If you and @Laurae2 are ok with it, we could do that for all the aliases (in a separate PR later)

**Comment by jameslamb** [comment]
> @jameslamb
> 
> > I'll update the language in #1143 to say "export callback functions" if that's alright.
> 
> I think that then this is another issue, even more feature request. But anyway, it's not related to pkgdown issue. I'll remove it from that list. Can you please create a separate issue for exporting callback functions?

Ok makes sense. I've created #2479 and attached it to #2302

File: R-package/R/lgb.cv.R (hunk 1 above)
Comment (StrikerRUS) [reply R-package/R/lgb.cv.R]: @jameslamb 
> we could do that for all the aliases (in a separate PR later)

I don't think we need a map for aliases of **all parameters**, since they are handled at cpp side. Only aliases of params which are used at R side should be there, I guess.

File: R-package/R/callback.R [hunk 2]
```diff
@@ -37,7 +37,12 @@ cb.reset.parameters <- function(new_params) {
 
     # Some parameters are not allowed to be changed,
     # since changing them would simply wreck some chaos
-    not_allowed <- c("num_class", "metric", "boosting_type")
+    not_allowed <- c(
+      "num_class"
+      , "metric"
+      , "boosting"
+      , .PARAMETER_ALIASES()[["boosting"]]
```
Comment (StrikerRUS) [inline R-package/R/callback.R]: I think `boosting` itself can be in returned values, so that only `.PARAMETER_ALIASES()[["boosting"]]` will be needed instead of current `"boosting", .PARAMETER_ALIASES()[["boosting"]]`.

**Comment by StrikerRUS** [comment]
@jameslamb I'm sorry, but could you please explain what is wrong with the following try to mimic Python solution on [pseudo?]R? https://www.diffchecker.com/LylbJ8EG

File: R-package/R/callback.R (hunk 2 above)
Comment (jameslamb) [reply R-package/R/callback.R]: you are right. If aliases are all in the cpp library there shouldn't a reason to care (in language wrappers) about what the "main" one is and what the "alias" ones are. I'll fix that.

**Comment by jameslamb** [comment]
> @jameslamb I'm sorry, but could you please explain what is wrong with the following try to mimic Python solution on [pseudo?]R? https://www.diffchecker.com/LylbJ8EG

I have never seen diffchecker.com before, super cool!

In my opinion, this code proposed in your link is going to be difficult to debug and reason about.

It's relying on a flag called `enabled` being present one level up the call stack (since it is not passed into `init()` and `params` being available an unknown number of levels up the call stack (since it is not passed in to `init()` or to `cb.early.stop()`.

So in my opinion while my proposal in this PR feels more tedious, it is also more _obvious_ and easier to debug / change.

I made that choice partially because I just feel it is a better design for the R package, but also partially because I am not so familiar with all of the R code package code that I could know the full implications of doing something like you proposed.

I will defer to @Laurae2 (who understand the internals of the R package more thoroughly) though. If @Laurae2 prefers something like the way this was handled in the Python package and pseudocode in your in link here, I'll happily change the approach from this PR and push the logic down into `cb.early.stop()`.

File: R-package/R/lgb.cv.R (hunk 1 above)
Comment (jameslamb) [reply R-package/R/lgb.cv.R]: Agree, sorry, I also mean "all aliases used in the R side"

**Comment by StrikerRUS** [comment]
@jameslamb Thanks a lot for your explanation!

**Comment by jameslamb** [comment]
I just rebased this to current `master`. ping @Laurae2 for a review

**Comment by jameslamb** [comment]
I just rebased this to latest `master` to catch up with changes over the last two weeks. It is ready for review.

**Comment by Laurae2** [comment]
@jameslamb As you pointed out there's a lot of things in common in `lgb.train` and `lgb.cv` which are still separated, you did the right things at the moment to duplicate the code.

Ideally in the future we should be able to merge all the duplicate code in one, but it requires rewriting the R package.

File: R-package/R/aliases.R [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# Central location for paramter aliases.
```
Comment (StrikerRUS) [inline R-package/R/aliases.R]:
```suggestion
# Central location for parameter aliases.
```

File: R-package/R/aliases.R [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+# See https://lightgbm.readthedocs.io/en/latest/Parameters.html#core-parameters
+
+# [description] List of respected parameter aliases. Wrapped in a function to take advantage of
+#               lazy evaluation (so it doesn't matter what order R sources files during installation).
+# [return] A named list, where each key is a main LightGBM parameter and  each value is a character
```
Comment (StrikerRUS) [inline R-package/R/aliases.R]:
```suggestion
# [return] A named list, where each key is a main LightGBM parameter and each value is a character
```

File: R-package/R/callback.R [hunk 5]
```diff
@@ -37,7 +37,12 @@ cb.reset.parameters <- function(new_params) {
 
     # Some parameters are not allowed to be changed,
     # since changing them would simply wreck some chaos
-    not_allowed <- c("num_class", "metric", "boosting_type")
+    not_allowed <- c(
+      "num_class"
+      , "metric"
```
Comment (StrikerRUS) [inline R-package/R/callback.R]: I think these params have their aliases, just like `boosting`.

File: R-package/R/callback.R [hunk 6]
```diff
@@ -37,7 +37,12 @@ cb.reset.parameters <- function(new_params) {
...
+      , "boosting"
```
Comment (StrikerRUS) [inline R-package/R/callback.R]: Refer to https://github.com/microsoft/LightGBM/pull/2443#discussion_r329503844.

File: R-package/R/callback.R (hunk 2 above)
Comment (jameslamb) [reply R-package/R/callback.R]: just changed this so now all valid names ("main" or otherwise) are returned

File: R-package/R/callback.R (hunk 5 above)
Comment (jameslamb) [reply R-package/R/callback.R]: added! good catch

File: R-package/R/callback.R (hunk 6 above)
Comment (jameslamb) [reply R-package/R/callback.R]: changed, thank you! Now it is

```r
not_allowed <- c(
      .PARAMETER_ALIASES()[["num_class"]]
      , .PARAMETER_ALIASES()[["metric"]]
      , .PARAMETER_ALIASES()[["boosting"]]
    )
```

**Comment by jameslamb** [comment]
I've rebased to `master` and implemented requested changes to handling of aliases.

File: R-package/R/callback.R (hunk 6 above)
Comment (StrikerRUS) [reply R-package/R/callback.R]: I think it's much better now! Thanks!

File: R-package/R/lgb.cv.R [hunk 7]
```diff
@@ -134,7 +134,7 @@ lgb.cv <- function(params = list(),
     begin_iteration <- predictor$current_iter() + 1
   }
   # Check for number of rounds passed as parameter - in case there are multiple ones, take only the first one
-  n_trees <- c("num_iterations", "num_iteration", "n_iter", "num_tree", "num_trees", "num_round", "num_rounds", "num_boost_round", "n_estimators")
+  n_rounds <- .PARAMETER_ALIASES()[["num_iterations"]]
   if (any(names(params) %in% n_trees)) {
     end_iteration <- begin_iteration + params[[which(names(params) %in% n_trees)[1]]] - 1
```
Comment (StrikerRUS) [inline R-package/R/lgb.cv.R]: I guess it should be either `n_trees`, or `n_rounds`.

Comment (jameslamb) [reply R-package/R/lgb.cv.R]: I am bad at writing code haha

Thank you, good catch. As soon as I do #2335 CI will catch this for us.

Comment (jameslamb) [reply R-package/R/lgb.cv.R]: ok this is fixed

**Comment by jameslamb** [comment]
Ok this has been rebased to `master` and I've fixed the `n_trees` vs. `n_rounds` thing

**Comment by jameslamb** [comment]
> LGTM, except two actually unresolved comments about grammar from [#2443 (review)](https://github.com/microsoft/LightGBM/pull/2443#pullrequestreview-306246905).

ah! ok fixing

**Comment by jameslamb** [comment]
Thank you for the reviews everyone! I am going to merge this once CI is done.

**Comment by jameslamb** [comment]
was getting this on `check-docs` task on Travis:

> 429: too many requests
> http://www.ozone3d.net/gpu_caps_viewer/

Restarting that task after waiting a few minutes seemed to solve it, but something to keep an eye on I guess.

**Comment by StrikerRUS** [comment]
@jameslamb 
> Restarting that task after waiting a few minutes seemed to solve it, but something to keep an eye on I guess.

Yeah, I see this error very often. And `docs` job is needed to be restarted quite often, that is why we keep it only at Travis where each collaborator has enough right to re-run it (https://github.com/microsoft/LightGBM/pull/1497#discussion_r201301028).
